File: src/main/java/com/pedigree/services/UndoRedoService.java
package com.pedigree.services;

import com.pedigree.editor.Command;
import com.pedigree.editor.CommandStack;
import com.pedigree.model.ProjectMetadata;

public class UndoRedoService {
    private final CommandStack commandStack = new CommandStack();
    private ProjectMetadata projectMetadata;

    public void bindProjectMetadata(ProjectMetadata meta) { this.projectMetadata = meta; }

    public void execute(Command command) { commandStack.execute(command); touch(); }
    public void undo() { commandStack.undo(); touch(); }
    public void redo() { commandStack.redo(); touch(); }
    public boolean canUndo() { return commandStack.canUndo(); }
    public boolean canRedo() { return commandStack.canRedo(); }
    public int getUndoCount() { return commandStack.getUndoSize(); }
    public int getRedoCount() { return commandStack.getRedoSize(); }

    private void touch() {
        if (projectMetadata != null) {
            projectMetadata.setModifiedAt(java.time.Instant.now());
        }
    }
}

File: src/main/kotlin/com/family/tree/services/UndoRedoService.kt
package com.family.tree.services

import com.family.tree.editor.Command
import com.family.tree.editor.CommandStack
import com.family.tree.model.ProjectMetadata
import java.time.Instant

class UndoRedoService {
    private val commandStack = CommandStack()
    private var projectMetadata: ProjectMetadata? = null

    fun bindProjectMetadata(meta: ProjectMetadata) { this.projectMetadata = meta }

    fun execute(command: Command) { commandStack.execute(command); touch() }
    fun undo() { commandStack.undo(); touch() }
    fun redo() { commandStack.redo(); touch() }
    fun canUndo(): Boolean = commandStack.canUndo()
    fun canRedo(): Boolean = commandStack.canRedo()
    fun getUndoCount(): Int = commandStack.getUndoSize()
    fun getRedoCount(): Int = commandStack.getRedoSize()

    private fun touch() {
        projectMetadata?.modifiedAt = Instant.now()
    }
}
